### PR TITLE
Controls 0, 1, and 2 - Vectorize Controls_config, Refactor Control Presets, and Use Undosys

### PR DESCRIPTION
**Pull request by z64555**
This PR combines branches `feature/controls`, `feature/controls1`, and `feature.controls2`. This was done because`feature/controls` breaks two features which are fixed in `feature/controls1` and `feature/controls2`, and the three should not be merged in individually.

<hr>

PR #2868 Controls0 - `feature/controls`

This changes the Controls_config array into an SCP_vector to allow scripting to add more controls as they wish, enabling mod devs to create new controls with scripting and not needing hardcode support in the future.  (Custom controls are still a ways off, however)

- [x] Bind mode behaves as it should
- [x] Search mode behaves as it should
- [x] Browse mode behaves as it should
- [x] Loading controlconfigdefaults with retail data behaves as it should
- [x] Loading controlconfigdefaults with mod data behaves as it should

<hr>

PR #2884 Controls 1 - `feature/controls1`

This re-factors control presets and adds a bunch of features.

* Hardcoded defaults are now a preset.
* Mod devs may override the hardcoded defaults with the controlconfigdefaults.tbl by naming a `#ControlConfigOverride` section as "default"
* Mod devs may add as many presets as they like with `#ControlConfigPreset` sections.  These have the same syntax and structure as `#ControlConfigOverride` but only add bindings, and start off as a blank slate.
* Added ability to change the name of controls as they show up in the menu using `$Text:`
* Added possibility of resumption, should a `$Bind Name:` not be found within the controls.

- [x] Reset to defaults should behave as it should
- [x] Repeat hitting of "Reset" button should cycle through all presets
- [x] controlconfigdefaults.tbl is able to override the default preset
- [x] controlconfigdefaults.tbl is able to make new presets with the #ControlConfigOverride section
- [x] controlconfigdefaults.tbl is able to make new presets with the #ControlConfigPreset section
- [x] controlconfigdefaults.tbl is able to relabel controls

<hr>

PR #2907 Controls 2 - `feature/controls2`

Introduces and improves the undo system from PR #1243 and uses it for the controls config menu to re-enable the undo feature.
- [x] Should be able to undo single buttons
- [x] Should be able to undo single axis bindings
- [x] Should be able to undo preset cycling
- [x] Should be able to undo multiple button/control changes (such as an entire wipe of the controls)
- [x] Should be able to do all vanilla functions normally

### CODE REVIEW

**Comment by z64555** [comment]
Noticed minor bug with drift.  Not sure if due to controller or something else.

**Comment by z64555** [comment]
Ok, I seem to have resolved all compile/build errors, so now that's all thats left is the automated tests.

I'm not sure why this is failing, as this group does minimal changes to the player files.

File: code/controlconfig/controlsconfig.cpp
Comment (asarium) [inline code/controlconfig/controlsconfig.cpp:182]: You can initialize the members directly as well:

```suggestion
	int key = -1;  // index of other control in conflict with this one
	int joy = -1;  // index of other control in conflict with this one
```
Then you do not need to explicitly declare the default constructor.

**Review by asarium — APPROVED** [review]
To be honest, this is too big to effectively review. I'm going to approve it but we should seriously think about including it in the upcoming release since there is a relatively large chance for a regression to happen.

**Comment by JohnAFernandez** [comment]
I'm with waiting.  Not really enough time to test and fix.  But merging right after stable gives us more than enough time to test and allows users to start taking advantage of these new features.

File: code/controlconfig/controlsconfig.cpp
Comment (z64555) [reply code/controlconfig/controlsconfig.cpp:182]: I think one of the CI's complained when I did that, most likely clang tidy.

**Review by Goober5000 — CHANGES_REQUESTED** [review]
Agreed, we should not merge this until after the stable release.  I'm putting this in the Request Changes state so that it's not merged accidentally.

**Review by Goober5000 — APPROVED** [review]

**Comment by z64555** [comment]
there's a number of issues that arose in testing this, and one of them is a freeze when pressing Clear All.

**Comment by z64555** [comment]
Latest commit should be green lit.  Ran through all checkboxes here.  We'll see how Controls3 will end up after this.

**Pull request merged**
